Lib/test/test_shelve.py

```diff
@@ -165,6 +166,277 @@ def test_default_protocol(self):
         with shelve.Shelf({}) as s:
             self.assertEqual(s._protocol, pickle.DEFAULT_PROTOCOL)
 
+    def test_custom_serializer_and_deserializer(self):
+        os.mkdir(self.dirname)
+        self.addCleanup(os_helper.rmtree, self.dirname)
+
+        def serializer(obj, protocol):
+            if isinstance(obj, (bytes, bytearray, str)):
+                if protocol == 5:
+                    return obj
+                else:
+                    return type(obj).__name__
+            elif isinstance(obj, array.array):
+                return obj.tobytes()
+            else:
+                raise TypeError(
+                    f"Unsupported type for serialization: {type(obj)}"
+                )
+
+        def deserializer(data):
+            if isinstance(data, (bytes, bytearray, str)):
+                return data.decode("utf-8")
+            elif isinstance(data, array.array):
+                return array.array("b", data)
+            else:
+                raise TypeError(
+                    f"Unsupported type for deserialization: {type(data)}"
+                )
+
+        for proto in range(pickle.HIGHEST_PROTOCOL + 1):
+            with self.subTest(proto=proto):
+                with shelve.open(
+                    self.fn,
+                    protocol=proto,
+                    serializer=serializer,
+                    deserializer=deserializer,
+                ) as s:
```

Comment:
You can have `with self.subTest(proto=proto), shelve.open(...) as s` to reduce the indentation level by 1.